File: book/petrol-pump/java/chapter4/section11/PresetAmountPump.java
package chapter4.section11;

import pump.*;
import chapter4.section4.LifeCycle;
import chapter4.section4.LifeCycle.End;
import chapter4.section6.AccumulatePulsesPump;
import chapter4.section7.Fill;
import chapter4.section7.ShowDollarsPump;
import chapter4.section8.NotifyPointOfSale;
import chapter4.section9.Keypad;
import nz.sodium.*;
import java.util.Optional;

public class PresetAmountPump implements Pump {
    public Outputs create(Inputs inputs) {
        StreamLoop<Fuel> sStart = new StreamLoop<>();
        Fill fi = new Fill(inputs.sClearSale.map(u -> Unit.UNIT),
                           inputs.sFuelPulses, inputs.calibration,
                           inputs.price1, inputs.price2, inputs.price3,
                           sStart);
        NotifyPointOfSale np = new NotifyPointOfSale(
                new LifeCycle(inputs.sNozzle1,
                              inputs.sNozzle2,
                              inputs.sNozzle3),
                inputs.sClearSale,
                fi);
        sStart.loop(np.sStart);
        CellLoop<Boolean> keypadActive = new CellLoop<>();
        Keypad ke = new Keypad(inputs.sKeypad,
                               inputs.sClearSale,
                               keypadActive);
        Preset pr = new Preset(ke.value,
                               fi,
                               np.fuelFlowing,
                               np.fillActive.map(o -> o.isPresent()));
        keypadActive.loop(pr.keypadActive);
        return new Outputs()
            .setDelivery(pr.delivery)
            .setSaleCostLCD(fi.dollarsDelivered.map(
                    q -> Formatters.formatSaleCost(q)))
            .setSaleQuantityLCD(fi.litersDelivered.map(
                    q -> Formatters.formatSaleQuantity(q)))
            .setPriceLCD1(ShowDollarsPump.priceLCD(np.fillActive, fi.price,
                    Fuel.ONE, inputs))
            .setPriceLCD2(ShowDollarsPump.priceLCD(np.fillActive, fi.price,
                    Fuel.TWO, inputs))
            .setPriceLCD3(ShowDollarsPump.priceLCD(np.fillActive, fi.price,
                    Fuel.THREE, inputs))
            .setSaleComplete(np.sSaleComplete)
            .setPresetLCD(ke.value.map(v ->
                Formatters.formatSaleCost((double)v)))
            .setBeep(np.sBeep.orElse(ke.sBeep));
    }
}

File: book/petrol-pump/c#/PetrolPump/Chapter4/Section11/PresetAmountPump.cs
using PetrolPump.Chapter4.Section4;
using PetrolPump.Chapter4.Section7;
using PetrolPump.Chapter4.Section8;
using PetrolPump.Chapter4.Section9;
using Sodium.Frp;

namespace PetrolPump.Chapter4.Section11
{
    public class PresetAmountPump : IPump
    {
        public Outputs Create(Inputs inputs)
        {
            StreamLoop<Fuel> sStart = new StreamLoop<Fuel>();
            Fill fi = new Fill(inputs.SClearSale,
                inputs.SFuelPulses, inputs.Calibration,
                inputs.Price1, inputs.Price2, inputs.Price3,
                sStart);
            NotifyPointOfSale np = new NotifyPointOfSale(
                new LifeCycle(inputs.SNozzle1,
                    inputs.SNozzle2,
                    inputs.SNozzle3),
                inputs.SClearSale,
                fi);
            sStart.Loop(np.SStart);
            CellLoop<bool> keypadActive = new CellLoop<bool>();
            Keypad ke = new Keypad(inputs.SKeypad,
                inputs.SClearSale,
                keypadActive);
            Preset pr = new Preset(ke.Value, fi, np.FuelFlowing);
            keypadActive.Loop(pr.KeypadActive);
            return new Outputs()
                .SetDelivery(pr.Delivery)
                .SetSaleCostLcd(fi.DollarsDelivered.Map(Formatters.FormatSaleCost))
                .SetSaleQuantityLcd(fi.LitersDelivered.Map(Formatters.FormatSaleQuantity))
                .SetPriceLcd1(ShowDollarsPump.PriceLcd(np.FillActive, fi.Price, Fuel.One, inputs))
                .SetPriceLcd2(ShowDollarsPump.PriceLcd(np.FillActive, fi.Price, Fuel.Two, inputs))
                .SetPriceLcd3(ShowDollarsPump.PriceLcd(np.FillActive, fi.Price, Fuel.Three, inputs))
                .SetSaleComplete(np.SSaleComplete)
                .SetPresetLcd(ke.Value.Map(Formatters.FormatPresetAmount))
                .SetBeep(np.SBeep.OrElse(ke.SBeep));
        }
    }
}